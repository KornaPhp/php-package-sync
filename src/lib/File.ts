/* eslint-disable no-unused-vars */

import { rmdirSync, existsSync, readFileSync, writeFileSync, unlinkSync, statSync } from 'fs';
import { basename, extname } from 'path';
import { isDirectory } from './helpers';

export enum FileType {
    FILE = 'file',
    DIRECTORY = 'directory',
}

export class File {
    protected data: string | null = null;
    protected type: FileType;

    constructor(protected fn: string, contents: string | null = null) {
        this.type = FileType.FILE;

        if (this.exists()) {
            this.type = isDirectory(fn) ? FileType.DIRECTORY : FileType.FILE;
        }

        this.data = contents;
    }

    static create(fn: string, contents: string | null = null) {
        return new File(fn, contents);
    }

    static read(fn: string) {
        return new File(fn)
            .load();
    }

    static contents(fn: string | null): string {
        if (!fn) {
            return '';
        }

        if (!fn.length) {
            return '';
        }

        return <string>File.create(fn).contents;
    }

    get filename() {
        return this.fn;
    }

    set filename(value: string) {
        this.fn = value;
    }

    get contents() {
        if (this.isDirectory()) {
            return '';
        }

        if (this.data === null) {
            this.load();
        }

        return this.data ?? '';
    }

    get sizeOnDisk() {
        if (this.isDirectory()) {
            return 0;
        }

        if (!this.exists()) {
            return 0;
        }

        const stat = statSync(this.fn);

        if (!stat.isFile()) {
            return 0;
        }

        return stat.size;
    }

    get basename() {
        return basename(this.fn);
    }

    get extension() {
        return extname(this.fn);
    }

    public isDirectory() {
        return this.type === FileType.DIRECTORY;
    }

    public isFile() {
        return this.type === FileType.FILE;
    }

    public load() {
        if (!this.isFile()) {
            return this;
        }

        this.data = readFileSync(this.fn, { encoding: 'utf-8' });

        return this;
    }

    public save() {
        return this.saveAs(this.fn);
    }

    public saveAs(fn: string) {
        if (this.isFile()) {
            writeFileSync(fn, this.data ?? '');
        }

        return this;
    }

    public delete() {
        if (this.isFile()) {
            unlinkSync(this.fn);
        }

        if (this.isDirectory()) {
            rmdirSync(this.fn);
        }

        return this;
    }

    public exists() {
        return existsSync(this.fn);
    }

    public setContents(data: string) {
        this.data = data;

        return this;
    }

    public processTemplate(repoName: string): string {
        return this.contents
            .replace(/:vendor_name/g, 'spatie')
            .replace(/:package_name/g, repoName)
            .replace(/author@domain\.com/g, 'freek@spatie.be')
            .replace(/:author_homepage/g, 'https://spatie.be/open-source/support-us');
    }
}
