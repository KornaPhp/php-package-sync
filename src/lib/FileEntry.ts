/* eslint-disable no-unused-vars */

import { ComparisonScoreRequirements } from '../types/ComparisonScoreRequirements';

export interface FileEntry {
    isFile: boolean;
    name: string;
    relativeName: string;
    filesize: number;
    shouldIgnore: boolean;
    shouldCompare: boolean;
    requiredScores: ComparisonScoreRequirements;
}

export class FileEntryArray extends Array<FileEntry> {
    constructor(entries: Array<any> | number) {
        if (Array.isArray(entries)) {
            super(...entries);
        } else {
            super(entries);
        }

        Object.setPrototypeOf(this, Object.create(FileEntryArray.prototype));
    }

    public containsEntry(entry: FileEntry): boolean {
        return !!(this.find(file => file.relativeName === entry.relativeName) || false);
    }

    public findByRelativeName(relativeName: string): FileEntry | null {
        return this.find(file => file.relativeName === relativeName) || null;
    }
}
