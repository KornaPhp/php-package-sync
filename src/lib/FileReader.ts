import { readFileSync } from 'fs';

export class FileReaderOld {
    static forCopy(fn: string): string {
        return readFileSync(fn, { encoding: 'utf-8' })
            .toString();
    }

    static forComparison(fn: string): string {
        return (
            readFileSync(fn, { encoding: 'utf-8' })
                .toString()
                // .replace(/\s{2,}/g, ' ')
                // .replace(/\n{2,}/g, '\n')
                .trim()
        );
    }

    static performReplacements(str: string, repoName: string): string {
        return str
            .replace(/:vendor_name/g, 'spatie')
            .replace(/:package_name/g, repoName)
            .replace(/author@domain\.com/g, 'freek@spatie.be')
            .replace(/:author_homepage/g, 'https://spatie.be/open-source/support-us');
    }

    static fromSkeleton(fn: string, repoName: string): string {
        return FileReaderOld.performReplacements(FileReaderOld.forComparison(fn), repoName);
    }

    static fromSkeletonForCopy(fn: string, repoName: string): string {
        return FileReaderOld.performReplacements(FileReaderOld.forCopy(fn), repoName);
        // .replace(/:vendor_name/g, 'spatie')
        // .replace(/:package_name/g, repoName)
        // .replace(/author@domain\.com/g, 'freek@spatie.be')
        // .replace(/:author_homepage/g, 'https://spatie.be/open-source/support-us');
    }
}

export class FileReader {
    protected data: string;

    constructor(protected fn: string) {
        this.data = readFileSync(fn, { encoding: 'utf-8' })
            .toString();
    }

    static create(fn: string) {
        return new FileReader(fn);
    }

    static contents(fn: string): string {
        return FileReader.create(fn).contents;
    }

    get filename() {
        return this.fn;
    }

    get contents() {
        return this.data;
    }

    public processTemplate(repoName: string): string {
        return this.data
            .replace(/:vendor_name/g, 'spatie')
            .replace(/:package_name/g, repoName)
            .replace(/author@domain\.com/g, 'freek@spatie.be')
            .replace(/:author_homepage/g, 'https://spatie.be/open-source/support-us');
    }
}
