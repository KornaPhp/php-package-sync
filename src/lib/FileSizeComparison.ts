/* eslint-disable no-unused-vars */

export class FileSizeComparison {
    static create(filesize1: number, filesize2: number) {
        return new FileSizeComparison(filesize1, filesize2);
    }

    constructor(public filesize1: number, public filesize2: number) {
        //
    }

    get difference(): number {
        return this.filesize1 - this.filesize2;
    }

    get percentDifference(): number {
        const diff = this.difference;

        return diff === 0 ? 0 : (diff / this.filesize1) * 100;
    }

    public differByPercentage(percentage = 15): boolean {
        const pctDecimal = percentage * 0.01;

        return this.difference > Math.round(this.filesize1 * pctDecimal);
    }

    public percentDifferenceForDisplay(padStart = 0, padEnd = 0): string {
        return (this.percentDifference.toFixed(2) + '%').padStart(padStart)
            .padEnd(padEnd);
    }
}

export function compareFileSizes(filesize1: number, filesize2: number): FileSizeComparison {
    return FileSizeComparison.create(filesize1, filesize2);
}
