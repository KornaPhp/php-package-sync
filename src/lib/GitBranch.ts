/* eslint-disable no-unused-vars */

import { runGitCommand } from './helpers';

export class GitBranch {
    constructor(public name: string) {
        //
    }

    public create() {
        runGitCommand(`branch ${this.name}`);

        return this;
    }

    public checkout(create = true) {
        runGitCommand(`checkout ${create ? '-b' : ''} ${this.name}`);

        return this;
    }

    public exists() {
        return !runGitCommand(`show-ref refs/heads/${this.name}`).empty;
    }

    public isCurrent() {
        return runGitCommand('branch --show-current')
            .equals(this.name);
    }
}
