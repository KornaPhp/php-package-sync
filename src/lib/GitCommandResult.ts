/* eslint-disable no-unused-vars */
import { SpawnSyncReturns } from 'child_process';

export class GitCommandResult {
    constructor(public result: SpawnSyncReturns<string>) {
        //console.log(result.pid);
    }

    get empty() {
        return this.result.output.length > 0;
    }

    get value() {
        return this.result.output.join('\n');
    }

    equals(str: string) {
        return str === this.value;
    }
}
