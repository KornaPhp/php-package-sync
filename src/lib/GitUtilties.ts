/* eslint-disable @typescript-eslint/ban-ts-comment */

import { existsSync } from 'fs';
import { basename } from 'path';
import { runCommand } from './helpers';

export class GitUtilties {
    public static displayStatusMessages = false;

    static pullRepo(name: string, path: string) {
        if (existsSync(path)) {
            if (GitUtilties.displayStatusMessages) {
                console.log(`* Updating repository '${basename(path)}'`);
            }
            runCommand('git', ['pull'], path);
        }
    }

    static cloneRepo(name: string, parentPath: string, cloneIntoDir: string | null = null) {
        // @ts-ignore
        cloneIntoDir = cloneIntoDir ?? name.split('/')
            .pop();

        const gitCloneUrl = `https://github.com/${name}.git`;

        if (!existsSync(parentPath + '/' + cloneIntoDir)) {
            if (GitUtilties.displayStatusMessages) {
                console.log(`* Cloning repository '${name}'`);
            }
            // @ts-ignore
            runCommand('git', ['clone', gitCloneUrl, cloneIntoDir], parentPath);
        }
    }
}
