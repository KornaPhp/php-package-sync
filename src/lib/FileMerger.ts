import { File } from './File';
import { LineMerger } from './LineMerger';

export class FileMerger {
    public merger: LineMerger;

    constructor() {
        this.merger = new LineMerger();
    }

    static create(): FileMerger {
        return new FileMerger();
    }

    public add(...filenames: string[]) {
        filenames.forEach(fn => {
            this.merger.add(File.contents(fn));
        });

        return this;
    }

    public merge(): string {
        return this.merger.merge()
            .join('\n');
    }

    public mergeAndSave(targetFile: string) {
        File.create(targetFile)
            .setContents(this.merge())
            .save();
    }
}
