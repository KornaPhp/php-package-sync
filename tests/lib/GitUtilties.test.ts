/* eslint-disable no-undef */

import { realpathSync } from 'fs';
import { GitUtilties } from '../../src/lib/GitUtilties';

let runCmdData: string[] = [];
const runCmd = function (cmd: string, args: string[], cwd: string) {
    const baseDir = realpathSync(__dirname + '/../..');

    runCmdData.push(`${cmd} ${args.join(' ')} ${cwd.split('/')
        .pop()}`.trim());

    runCmdData = runCmdData.map(line => line.replace(baseDir, '.'));
};

beforeEach(() => {
    runCmdData = [];
    GitUtilties.runCmd = runCmd;
});

it("does not run any git commands to pull a repository when the target dir doesn't exist", () => {
    GitUtilties.pullRepo('vendorname/some-repo-name', 'temp');

    expect(runCmdData)
        .toHaveLength(0);
});

it('runs a git command to pull a repository when the target dir exists', () => {
    GitUtilties.pullRepo('vendorname/some-repo-name', __dirname);

    expect(runCmdData)
        .toMatchSnapshot();
});

it("runs the correct git command to clone a repository when the target dir doesn't exist", () => {
    GitUtilties.cloneRepo('vendorname/some-repo-name', 'temp', 'target-dir');

    expect(runCmdData)
        .toMatchSnapshot();
    expect(runCmdData.join('')
        .startsWith('git clone'))
        .toBeTruthy();
});

it("doesn't run a git command to clone a repository when the target dir exists", () => {
    GitUtilties.cloneRepo('vendorname/some-repo-name', realpathSync(__dirname + '/../data'), 'test-package-2');

    expect(runCmdData)
        .toHaveLength(0);
});
