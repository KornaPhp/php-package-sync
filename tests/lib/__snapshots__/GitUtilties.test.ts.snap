// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runs a git command to pull a repository when the target dir exists 1`] = `
Array [
  "git pull lib",
]
`;

exports[`runs the correct git command to clone a repository when the target dir doesn't exist 1`] = `
Array [
  "git clone https://github.com/vendorname/some-repo-name.git target-dir temp",
]
`;
